src/Interpreters/InterpreterDropQuery.cpp

```diff
@@ -202,7 +206,15 @@ BlockIO InterpreterDropQuery::executeToTableImpl(ContextPtr context_, ASTDropQue
 
             table->checkTableCanBeDropped();
 
-            auto table_lock = table->lockExclusively(context_->getCurrentQueryId(), context_->getSettingsRef().lock_acquire_timeout);
+            TableExclusiveLockHolder table_lock;
+            /// We don't need this lock for ReplicatedMergeTree
+            if (!table->supportsReplication())
+            {
+                /// And for simple MergeTree we can stop merges before acquiring the lock
+                auto merges_blocker = table->getActionLock(ActionLocks::PartsMerge);
```

Comment:
This code looks incorrect. The locks go out of scope immediately, so truncate() is being called without holding them, right?

Reply:
Yep, it's incorrect, will be fixed in https://github.com/ClickHouse/ClickHouse/pull/41145